Add Letter-spacing and enable this for site title and site tagline

carolinan opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
Adds a new letter-spacing control to the typography panel.
Enables the control for site title and site tagline blocks.

Partial for https://github.com/WordPress/gutenberg/issues/30918, https://github.com/WordPress/gutenberg/issues/30921

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

I have tested three blocks manually in the site editor and block editor:
Site title and tagline with the control.
Site title and paragraph with defaults declared in theme.json.

To enable letter-spacing in theme.json:
```
settings > defaults > typography > "customLetterSpacing": true,
styles > block > typography > "letterSpacing": "10px"

```

## Screenshots <!-- if applicable -->

Block editor:
![Letter-spacing control in site editor](https://user-images.githubusercontent.com/7422055/115828276-c5809300-a40d-11eb-96c9-0120c59c2915.png)

Site editor: ❓ **Help wanted. I am not sure why the control looks different in the site editor.**
![Letter-spacing control in site editor](https://user-images.githubusercontent.com/7422055/115828039-70448180-a40d-11eb-9cac-bf7e0ff75c90.png)

Global styles:
![global styles typography panel](https://user-images.githubusercontent.com/7422055/115845569-45fcbf00-a421-11eb-99ed-b55d510588fd.png)


Front:
![Site title and tagline with letter spacing](https://user-images.githubusercontent.com/7422055/115827936-4ab77800-a40d-11eb-9db7-b39d6efd9e90.png)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
New control

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

carolinan committed:
TRY: Letter-spacing

carolinan committed:
Linting

carolinan committed:
Change to VisualLabel

carolinan committed:
More linting

carolinan committed:
Update the site editor typography panel

carolinan committed:
Linting

carolinan committed:
This is embarassing, I really need Prettier to work locally

ockham commented on lib/class-wp-theme-json.php:
```diff
@@ -92,6 +92,7 @@ class WP_Theme_JSON {
 			'lineHeight'     => null,
 			'textDecoration' => null,
 			'textTransform'  => null,
+			'letterSpacing'  => null,
```
Small nitpick -- it looks like this list is alphabetized. Should we maybe put `letterSpacing` before `lineHeight`?

ockham commented on lib/class-wp-theme-json.php:
```diff
@@ -269,6 +271,9 @@ class WP_Theme_JSON {
 		'text-transform'           => array(
 			'value' => array( 'typography', 'textTransform' ),
 		),
+		'letter-spacing'           => array(
+			'value' => array( 'typography', 'letterSpacing' ),
+		),
```
Same as https://github.com/WordPress/gutenberg/pull/31118/files#r625025653

ockham reviewed (commented):
Thanks @carolinan, this is looking good and working well in my testing! 👍  From a technical point of view, this LGTM.

It might be nice to get a designer to have a quick look at this and maybe help polish it a bit -- I think I saw a convo between @mtias, @jasmussen, and @jameskoster the other day about styling typography related controls. (One point I remember was that we now have some controls that do have a 'Reset' button while others don't, and it seems a bit arbitrary.) For reference, here's an issue with a few mockups: #27331.

jasmussen commented:
Thanks for the ping. Yes, we really need a progressive-reveal system as described in #27331, especially important for somewhat exotic properties like this one which nevertheless need to be available:

<img width="1670" alt="mockup" src="https://user-images.githubusercontent.com/1204802/116884395-fd19e700-ac26-11eb-835b-501876ad0c7f.png">

The prominence in that mockup also weights the letter spacing as "half a column", rather than a full row, but that does seem like it's made easier by the fact that simply unchecking the property in that mockup also serves as a reset.

What is the closest we can get to that mockup, in absence of the progressive-reveal system?

carolinan committed:
Sync with trunk (#31427)

carolinan commented:
I did not realize that would ping everyone! I'm sorry!

ockham commented:
> I did not realize that would ping everyone! I'm sorry!

It has definitely happened to me! Looks like the sync didn't quite work out (and the PR diff now includes 895 files 😅 ). Can you try reverting f2046849ac54f51482272e3d7299f45856b0240e?

ockham commented:
Thanks for weighing in, @jasmussen !

> What is the closest we can get to that mockup, in absence of the progressive-reveal system?

 Good question; not sure if we have a grid/column system for sidebar controls (yet) 🤔 

Barring that, it'd be interesting to know whether
- we should use `UnitControl` -- which allows to select a unit (`px`, `em`, `rem`), whereas other controls (e.g. Line Height) don't.
- we should include a Reset button.

carolinan committed:
Revert "Sync with trunk (#31427)" (#31428)

This reverts commit f2046849ac54f51482272e3d7299f45856b0240e.

jasmussen commented:
Threading the needle is tricky on this one. If we have good confidence that the specific _implementation_ is good, and that mostly the same can be used once we _do_ get a chance to roll out the component system, then perhaps in the mean time it's okay for this feature to be added only to the global styles system. We want to be very careful in surfacing it for post-editor facing blocks until we have something that's simpler and more userfriendly.

Another option is to not surface it with a user-facing UI at all, and have it be entirely theme.json specific...

carolinan commented:
I don't think I will manage to sync this with trunk without creating a new PR. Because the G2 components where removed, this PR can't be built anymore.  So, help wanted 😬 

----

Without the blocks supporting the letter spacing control, is it still possible to create block patterns with different letter spacing (picture different options for header template parts), or would it only support the default from theme.json?

ockham commented:
> I don't think I will manage to sync this with trunk without creating a new PR. Because the G2 components where removed, this PR can't be built anymore. So, help wanted 😬

Sure, I'll have a look. Hopefully I can just rebase it locally 🙂 

> Without the blocks supporting the letter spacing control, is it still possible to create block patterns with different letter spacing (picture different options for header template parts), or would it only support the default from theme.json?

If block support isn't enabled for the letter spacing feature, I don't think it'll be possible to create block patterns to use it. However, if I'm reading Joen correctly, we might consider keeping the "backend" parts and just remove the UI parts. 🤔

ockham commented:
> Threading the needle is tricky on this one. If we have good confidence that the specific _implementation_ is good, and that mostly the same can be used once we _do_ get a chance to roll out the component system, then perhaps in the mean time it's okay for this feature to be added only to the global styles system. We want to be very careful in surfacing it for post-editor facing blocks until we have something that's simpler and more userfriendly.

I think that the implementation is overall fine (and a straight-forward mapping of the corresponding CSS property) -- so we might just proceed with this as-is, and 'simply' not enable it for the post editor blocks.

Specifically, the Reset button is definitely something we can revisit later. As for the `UnitControl` -- we might want to decide early on whether or not we want to allow the user to set the unit (when they can't do that for e.g. line height). While it'll still be possible to change this later, we'd likely need a block migration to transform the attribute, and it'd be nice if we could avoid that.

I was mentioning the difference to the line height control (which doesn't include a unit), but I realize now that [line height](https://developer.mozilla.org/en-US/docs/Web/CSS/line-height) uses a relative unit (relative to the font size), which [letter spacing](https://developer.mozilla.org/en-US/docs/Web/CSS/letter-spacing) doesn't support 🤔

ockham commented:
> > I don't think I will manage to sync this with trunk without creating a new PR. Because the G2 components where removed, this PR can't be built anymore. So, help wanted 😬
> 
> Sure, I'll have a look. Hopefully I can just rebase it locally 🙂

Update, I just finished running

```
npm run distclean
npm i
npm run build
```

for this branch locally, and it seems to build fine -- no rebase needed. (CI checks going green seem to somehow confirm that.)
@carolinan What kind of build error are you seeing? Does the above sequence of commands fix the issue for you?

carolinan commented:
These are the types of errors I get when I try to build it locally:
```
packages/components/src/ui/view/component.js:4:26 - error TS2307: Cannot find module '@wp-g2/styles' or its corresponding type declarations.
4 import { BaseView } from '@wp-g2/styles';
```

And when I use `npm run distclean`, rimraf is not found, even though it is in the node_modules folder. Will try to install rimraf globally and check again.
```
> rimraf node_modules packages/*/node_modules

The system cannot find the path specified.
```

carolinan committed:
try to solve merge conflict

carolinan committed:
Merge branch 'trunk' into try/letter-spacing

aristath commented:
@carolinan I had the same issue on trunk a couple of weeks ago and it turned out to be a node version error. Downgrading my node to v14.16.1 fixed the issue on my end (used nvm). Please note that after switching node versions you'll have to `npm install` again.

carolinan commented:
It's some sort of magic ;), because after restarting my computer this morning, I can build the PR without errors.

carolinan commented:
I have solved the nit picks :) but we still need a decision about the units and reset button.

jasmussen commented:
Units feel important to letter spacing, in a way they aren't for line-height. 

If emptying (selecting the number and deleting it) and saving accomplishes the same as pressing a "Clear" button, we should remove the clear button. We will revisit this anyway in the future.

carolinan committed:
Remove the reset button

Remove the reset button.
Add style.scss in order to position the visible label.
Remove the limit on the negative value.

carolinan commented:
https://github.com/WordPress/gutenberg/pull/31118/commits/cb62a80a6636c6b131c7ddc63655831fe5ca9fef  removes the reset button and the limit on the negative value.
Adds a .scss file so that the visible label can be positioned at a distance that is consistent with other labels in the typography panel.



New screenshot:
![Letter-spacing control without reset button](https://user-images.githubusercontent.com/7422055/117133598-c073fa00-ada4-11eb-8ad9-dd2f09901974.png)

carolinan committed:
Merge branch 'trunk' into try/letter-spacing

carolinan committed:
linting

ockham commented on packages/block-editor/src/components/letter-spacing-control/index.js:
```diff
@@ -0,0 +1,54 @@
+ */
+export default function LetterSpacingControl( { value, onChange } ) {
+	return (
+		<BaseControl className="block-editor-letter-spacing-control">
+			<BaseControl.VisualLabel>
+				{ __( 'Letter-spacing' ) }
+			</BaseControl.VisualLabel>
+			<UnitControl
+				aria-label={ __( 'Letter-spacing' ) }
+				value={ value }
+				__unstableInputWidth="60px"
+				units={ CSS_UNITS }
+				onChange={ onChange }
+			/>
+		</BaseControl>
```
`UnitControl` supports a `label` prop; can we use that instead of `BaseControl` here? I think it'd make the code more semantic; and I think it's overall aligned with the component's intended usage. Furthermore, I think it'd allow us to drop `aria-label` (which I think uses the value of `label` as a fallback):

```suggestion
		<UnitControl
			aria-label={ __( 'Letter-spacing' ) }
			label={ __( 'Letter-spacing' ) }
			value={ value }
			__unstableInputWidth="60px"
			units={ CSS_UNITS }
			onChange={ onChange }
		/>
```

This will then also allow us to drop `packages/block-editor/src/components/letter-spacing-control/style.scss` 🙂

carolinan replied:
I tried this when I removed the flex earlier,
but when I remove the .scss and the BaseControl, and add the label to UnitControl, there is a `margin: 0` added to the label and it is not positioned at the same distance as the other visible labels.

```
.css-1dn6kmm-View-BaseLabel.css-1dn6kmm-View-BaseLabel.css-1dn6kmm-View-BaseLabel {
    box-sizing: border-box;
    color: currentcolor;
    display: block;
    margin: 0px;
    max-width: 100%;
    z-index: 1;
    padding-top: 0px;
    padding-bottom: 4px;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}
```
I also tried a version without the BaseControl but with a .scss file but I was not able to override this CSS.

carolinan replied:
The label for the font family control is also not at the same distance as the other controls in the typography panel, so the distances are already inconsistent. Which solution do we prefer?

![image](https://user-images.githubusercontent.com/7422055/117686268-fe16bf80-b1b6-11eb-8c42-c6a685e6f005.png)

ockham replied:
Ah right, thanks for explaining!

> The label for the font family control is also not at the same distance as the other controls in the typography panel, so the distances are already inconsistent. Which solution do we prefer?

I'm still leaning toward the `label` prop on `UnitControl`. I'd consider the missing margin a bug, and I'd prefer getting that fixed (especially as we're expecting the entire typography panel to get a design overhaul) over working around it by directly using `BaseControl`.

ockham reviewed (commented):
Thanks @carolinan, looking pretty good now! I have [one more suggestion](https://github.com/WordPress/gutenberg/pull/31118/files#r629287032) 😄

carolinan committed:
Remove BaseControl

ockham reviewed (approved):
Thank, LGTM! :shipit:

carolinan commented:
`useEditorFeature` needs to be replaced with `useSetting`

carolinan commented:
Awaiting https://github.com/WordPress/gutenberg/pull/31782

ockham commented:
@carolinan Are you sure you want to wait for #31782? It's very considerate of you, but I don't think it's a strict requirement -- that PR is still only in draft state, so it might take a while before it reaches approval state, and it might become progressively harder to rebase _this_ PR.

aristath commented:
I think it would be best to proceed without #31782. That PR will take a long time to be ready - if it ever is.

carolinan committed:
Replace useEditorFeature

carolinan committed:
Merge branch 'trunk' into try/letter-spacing

carolinan committed:
linting

carolinan committed:
Merge branch 'trunk' into try/letter-spacing

carolinan committed:
Merge branch 'trunk' into try/letter-spacing

carolinan committed:
try to resolve merge conflict

carolinan committed:
Merge branch 'trunk' into try/letter-spacing

noisysocks commented:
Hey @carolinan do you remember if it was a conscious decision to not add _Letter spacing_ to the Global Styles typography panel?

carolinan commented:
It was not, I don't think it even crossed my mind to add it there too :)
